### PR TITLE
CPLAT-357: React 16 from NPM and Bundled with Webpack

### PR DESCRIPTION
**Pull request by kealjones-wk**
### Motivation
We want to start using React 16 in react.dart, Starting with React 16, React no longer comes bundled with certain pieces like `React.createClass`, `React.PropTypes`, or `ShallowRenderer`.

### Changes
- Pull in React as a module and build custom bundles using webpack
    - Set up JS proxy component class using Babel ES6 classes.
    - Pull in the create-react-class module
    - Pull in `PropTypes` module for usage in context
    - Pull in and export React test utils (dev/testing bundles only)
    - Pull in and export ShallowRenderer to support consumer tests (dev/testing bundles only)
    - Turn on source maps for prod builds.
- Mirror existing JS bundles exposed in lib.
- This PR does not implement/expose any new react 16 features, it only gets the package ready too. 

### Notes
- New Context is not implemented in this ticket, currently supporting Legacy Context. (will probably update in another ticket/pr)
- ReactComponent.isMounted() has been removed. external libraries such as over_react_test rely on this in a few places `jacket.getInstance().isMounted()`
- React.createClass is not needed by this branch but external libraries such as over_react_test may rely on it, so it is only needed if we want to support external usage of it.
- React TestUtils has removed `SimulateNative`.

### Testing/QA
- [ ] CI Passes
- [ ] Run `webdev serve` navigate to http://localhost:8080/geocodes/geocodes.html and in the console type `React.version` and it should output "16.8.2"

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -5,4 +5,8 @@ pubspec.lock
 .idea
 .pub/
 .packages
+<<<<<<< HEAD
```
Comment (joelleibow-wf) [inline .gitignore]: Looks like you committed with a merge conflict, maybe?

File: .vscode/settings.json [hunk 2]
```diff
@@ -0,0 +1,3 @@
+{
+    "dart.sdkPath": "/usr/local/Cellar/dart/1.24.3/libexec"
```
Comment (joelleibow-wf) [inline .vscode/settings.json]: Do we want to have these settings in the to repo?

File: js_src/_dart_helpers.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
+/**
```
Comment (joelleibow-wf) [inline js_src/_dart_helpers.js]: Does this file get transpiled?

File: .gitignore (hunk 1 above)
Comment (kealjones-wk) [reply .gitignore]: no joel dont you know... thats the cool new .gitignore that will auto solve all merge conflicts.

File: .vscode/settings.json (hunk 2 above)
Comment (kealjones-wk) [reply .vscode/settings.json]: nope

File: js_src/_dart_helpers.js (hunk 3 above)
Comment (kealjones-wk) [reply js_src/_dart_helpers.js]: it goes through babel and webpacked

File: js_src/_dart_helpers.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+          return dartInteropStatics.handleGetChildContext(this.dartComponent);
+        };
+      }
+
+      if (contextKeys && contextKeys.length !== 0) {
+        ReactDartComponent.contextTypes = {};
+        for (var i = 0; i < contextKeys.length; i++) {
+          ReactDartComponent.contextTypes[contextKeys[i]] = React.PropTypes.object;
+        }
+      }
```
Comment (greglittlefield-wf) [inline js_src/_dart_helpers.js]: Per our conversation this morning, we can just remove this, right?

File: js_src/react_dom_dev.js [hunk 5]
```diff
@@ -0,0 +1,10 @@
+const ShallowRenderer = require('react-test-renderer/shallow');
+const TestUtils = require('react-dom/test-utils');
+
+
+if (!window.React.addons) {
+    window.React.addons = {};
+}
+window.React.addons.TestUtils = TestUtils;
+window.ReactShallowRenderer = ShallowRenderer;
```
Comment (greglittlefield-wf) [inline js_src/react_dom_dev.js]: Should we preserve this path under test utils? 🤷‍♂️

File: lib/react.js [hunk 6]
```diff
@@ -1,4394 +1,301 @@
...
...dContextKeys && childContextKeys.length !== 0) {\n        ReactDartComponent.childContextTypes = {};\n        for (var i = 0; i < childContextKeys.length; i++) {\n          ReactDartComponent.childContextTypes[childContextKeys[i]] = React.PropTypes.object;\n        }\n        // Only declare this when `childContextKeys` is non-empty to avoid unnecessarily\n        // creating interop context objects for components that won't use it.\n        ReactDartComponent.prototype['getChildContext'] = function() {\n          return dartInteropStatics.handleGetChildContext(this.dartComponent);\n        };\n      }\n\n      if (contextKeys && contextKeys.length !== 0) {\n        ReactDartComponent.contextTypes = {};\n        for (var i = 0; i < contextKeys.length; i++) {\n          ReactDartComponent.contextTypes[contextKeys[i]] = React.PropTypes.object;\n        }\n      }\n\n  return ReactDartComponent;\n}\n\nfunction _markChildValidated(child) {\n  const store = child._store;\n  if (store) store.validated = true;\n}\n\nmodule.exports = {\n  _getProperty,\n  _setProperty,\n  _createReactDartComponentClass,\n  _markChildValidated,\n};\n\n\n//# sourceURL=webpack:///./js_src/_dart_helpers.js?");
```
Comment (greglittlefield-wf) [inline lib/react.js]: We can't use `eval` in any of our deployed environments due to CSP.

While most of the time we'll be deploying prod React JS, there are cases where the development version would be useful to pull in. Can we configure webpack to not use eval?

File: test/ReactSetStateTestComponent.js [hunk 7]
```diff
@@ -10,68 +10,67 @@ function getNonUpdatingSetStateLifeCycleCalls() {
...
-    return {shouldUpdate: true};
-  },
-
-  getInitialState: function() {
-    return {counter: 0};
-  },
+class ReactSetStateTestComponent extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {counter: 1};
```
Comment (greglittlefield-wf) [inline test/ReactSetStateTestComponent.js]: Why was this changed from `0` to `1`?

File: test/factory/common_factory_tests.dart [hunk 8]
```diff
@@ -209,12 +210,12 @@ void _childKeyWarningTests(Function factory) {
...
-      expect(
-          consoleErrorMessage,
-          contains(
-              'Each child in an array or iterator should have a unique "key" prop.'));
+      expect(consoleErrorMessage,
+          contains('Each child in a list should have a unique "key" prop.'));
     });
 
+    /*
+    /// This might not be a thing in react16.
```
Comment (greglittlefield-wf) [inline test/factory/common_factory_tests.dart]: Can we address this before the PR merges?

File: js_src/_dart_helpers.js (hunk 4 above)
Comment (kealjones-wk) [reply js_src/_dart_helpers.js]: I was planning on doing that in the context Ticket

File: js_src/react_dom_dev.js (hunk 5 above)
Comment (kealjones-wk) [reply js_src/react_dom_dev.js]: Hmm nice catch, ill look into this.

File: lib/react.js (hunk 6 above)
Comment (kealjones-wk) [reply lib/react.js]: Ill check it out.

File: test/ReactSetStateTestComponent.js (hunk 7 above)
Comment (kealjones-wk) [reply test/ReactSetStateTestComponent.js]: I never finished this portion will do that now. :P

File: test/factory/common_factory_tests.dart (hunk 8 above)
Comment (kealjones-wk) [reply test/factory/common_factory_tests.dart]: Updated to check that it does NOT warn as this is now a valid return.

File: webpack.config.js [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        {
+          output: {
+            path: outputPath,
+            filename: outputFilename
+          },
+          entry: path.resolve(inputPath, entryFilename),
+          plugins: isProduction ? prodPlugins : devPlugins,
+          mode: isProduction ? "production" : "development",
+          externals: [{ window: "window" }],
+          devtool: isProduction ? false : 'inline-source-map'
```
Comment (greglittlefield-wf) [inline webpack.config.js]: Looking at [the documentation](https://webpack.js.org/configuration/devtool/#production), it seems like `source-map` might be best for all cases:
```suggestion
          devtool: 'source-map'
```

Advantages over current setup:
- development: source maps are a separate file, so they don't have to be loaded when you're not actively using them
- prod: source maps are available for prod builds, allowing prod stack traces to be mapped properly and allowing for easier prod debugging

File: js_src/react_dom_dev.js (hunk 5 above)
Comment (greglittlefield-wf) [reply js_src/react_dom_dev.js]: Do we still need this as a top-level export as well?

Comment (kealjones-wk) [reply js_src/react_dom_dev.js]: apparently not.

File: webpack.config.js (hunk 9 above)
Comment (kealjones-wk) [reply webpack.config.js]: nice
